include/deal.II/base/mpi.h

```diff
@@ -708,6 +709,416 @@ namespace Utilities
+       * the request and send an answer.
+       */
+      void
+      process_requests(int index);
+
+      /**
+       * Start to send all requests via ISend and post IRecvs for the incoming
+       * answer messages.
+       */
+      unsigned int
+      start_communication();
+
+      /**
+       * After all answers have been exchanged, the MPI data structures can be
+       * freed and the received answers can be processed.
+       */
+      void
+      clean_up_and_end_communication();
+    };
+
+    /**
+     * A class which delegates its task to other ConsensusAlgorithm
+     * implementations depending on the number of processes in the
+     * MPI communicator. It uses for small number of processes
+     * ConsensusAlgorithm_PEX and for large number of processes
+     * ConsensusAlgorithm_NBX. The threshold depends if the program is
+     * compiled in debug or release mode.
+     *
+     * @author Peter Munch, 2019
+     */
+    template <typename T1, typename T2>
+    class ConsensusAlgorithmSelector : public ConsensusAlgorithm<T1, T2>
+    {
+    public:
+      /**
+       * Constructor.
+       *
+       * @param process Process to be run during consensus algorithm.
+       * @param comm MPI Communicator.
+       */
+      ConsensusAlgorithmSelector(ConsensusAlgorithmProcess<T1, T2> &process,
+                                 const MPI_Comm &                   comm);
+
+      /**
+       * Destructor.
+       */
+      virtual ~ConsensusAlgorithmSelector();
+
+      /**
+       * Run consensus algorithm. The function call is delegated to another
+       * ConsensusAlgorithm implementation.
+       */
+      void
+      run() override;
+
+    private:
+      // Pointer to the actual ConsensusAlgorithm implementation.
+      std::shared_ptr<ConsensusAlgorithm<T1, T2>> consensus_algo;
+    };
+
+    /**
+     * Given a partitioned index set space, compute the owning MPI process rank
+     * of each element of a second index set according to the partitioned index
+     * set. A natural usage of this function is to compute for each ghosted
+     * degree of freedom the MPI rank of the process owning that index.
+     *
+     * One might think: "But we know which rank a ghost DoF belongs to based on
+     * the subdomain id of the cell it is on". But this heuristic fails for DoFs
+     * on interfaces between ghost cells with different subdomain_ids, or
+     * between a ghost cell and an artificial cell."
+     *
+     * The first argument passed to this function, @p owned_indices, must
+     * uniquely partition an index space between all processes.
+     * Otherwise, there are no limitations on this argument: In particular,
+     * there is no need in partitioning
+     * the index space into contiguous subsets. Furthermore, there are no
+     * limitations
+     * on the second index set @p indices_to_look_up as long as the size matches
+     * the first one. It can be chosen arbitrary and independently on each
```

Comment:
```suggestion
     * the first one. It can be chosen arbitrarily and independently on each
```

Reply:
In case the list of index sets can be injective only, you want that the ghosted sets are contained in the union of all locally owned parts on all processors.